[BUG] pinned_memory_resource (host MR) fails with CCCL 3.1.0+

**Describe the bug**
PR #2017 added build-time support for CCCL 3.1.0. In downstream testing, I found that cuVS failed to compile:

```
/home/coder/cuvs/cpp/src/core/c_api.cpp: In lambda function:
/home/coder/cuvs/cpp/src/core/c_api.cpp:171:31: error: no matching function for call to 'rmm::mr::pinned_memory_resource::allocate(const size_t&)'
  171 |     *ptr = pinned_mr->allocate(bytes);
      |            ~~~~~~~~~~~~~~~~~~~^~~~~~~
In file included from /home/coder/cuvs/cpp/src/core/c_api.cpp:33:
/home/coder/rmm/cpp/include/rmm/mr/host/pinned_memory_resource.hpp:166:9: note: candidate: 'void* rmm::mr::pinned_memory_resource::allocate(rmm::cuda_stream_view, std::size_t, std::size_t)'
  166 |   void* allocate(cuda_stream_view stream, std::size_t bytes, std::size_t alignment)
      |         ^~~~~~~~
/home/coder/rmm/cpp/include/rmm/mr/host/pinned_memory_resource.hpp:166:9: note:   candidate expects 3 arguments, 1 provided
/home/coder/cuvs/cpp/src/core/c_api.cpp: In lambda function:
/home/coder/cuvs/cpp/src/core/c_api.cpp:177:70: error: no matching function for call to 'rmm::mr::pinned_memory_resource::deallocate(void* const&, const size_t&)'
  177 |   return cuvs::core::translate_exceptions([=] { pinned_mr->deallocate(ptr, bytes); });
      |                                                 ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/home/coder/rmm/cpp/include/rmm/mr/host/pinned_memory_resource.hpp:180:8: note: candidate: 'void rmm::mr::pinned_memory_resource::deallocate(rmm::cuda_stream_view, void*, std::size_t, std::size_t)'
  180 |   void deallocate(cuda_stream_view stream, void* ptr, std::size_t bytes, std::size_t alignment)
      |        ^~~~~~~~~~
/home/coder/rmm/cpp/include/rmm/mr/host/pinned_memory_resource.hpp:180:8: note:   candidate expects 4 arguments, 2 provided
```

After some investigation, I found that the changes in this block triggered C++ inheritance name hiding.

https://github.com/rapidsai/rmm/blob/278d5596227df7ba91f59b056f1dfe34d6af68f7/cpp/include/rmm/mr/host/pinned_memory_resource.hpp#L152-L185

The `pinned_memory_resource` defines new `allocate` overloads to be compatible with CCCL 3.1.0+ memory resource concepts. However, the previous `allocate(bytes)` is hidden.

This should have been caught in RMM tests. I am working on a PR that fixes the bug (with `using host_memory_resource::allocate` to explicitly allow the parent's overloads) and add tests that will catch this for the future.

--- a/cpp/include/rmm/mr/host/pinned_memory_resource.hpp
+++ b/cpp/include/rmm/mr/host/pinned_memory_resource.hpp
@@ -152,6 +152,12 @@ class pinned_memory_resource final : public host_memory_resource {
 #if CCCL_MAJOR_VERSION > 3 || (CCCL_MAJOR_VERSION == 3 && CCCL_MINOR_VERSION >= 1)
 
  public:
+  // Explicitly inherit the allocate and deallocate functions from the host_memory_resource class.
+  // Due to inheritance and name hiding rules, we need to declare these with "using" when we
+  // override allocate and deallocate for CCCL 3.1.0+ compatibility.
+  using host_memory_resource::allocate;
+  using host_memory_resource::deallocate;
+
   /**
    * @brief Pretend to support the allocate_async interface, falling back to stream 0
    *

--- a/cpp/tests/CMakeLists.txt
+++ b/cpp/tests/CMakeLists.txt
@@ -183,6 +183,9 @@ ConfigureTest(LIMITING_TEST mr/device/limiting_mr_tests.cpp)
 # host mr_ref tests
 ConfigureTest(HOST_MR_REF_TEST mr/host/mr_ref_tests.cpp)
 
+# pinned mr tests
+ConfigureTest(PINNED_MR_TEST mr/host/pinned_mr_tests.cpp)
+
 # pinned pool mr tests
 ConfigureTest(PINNED_POOL_MR_TEST mr/host/pinned_pool_mr_tests.cpp)
 

--- a/cpp/tests/mr/host/pinned_mr_tests.cpp
+++ b/cpp/tests/mr/host/pinned_mr_tests.cpp
@@ -1,0 +1,46 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <rmm/mr/host/pinned_memory_resource.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <gtest/gtest.h>
+
+namespace rmm::test {
+namespace {
+
+// Returns true if a pointer points to pinned host memory.
+inline bool is_pinned_memory(void* ptr)
+{
+  cudaPointerAttributes attributes{};
+  if (cudaSuccess != cudaPointerGetAttributes(&attributes, ptr)) { return false; }
+  return attributes.type == cudaMemoryTypeHost;
+}
+
+}  // namespace
+
+// Issue 2057: Ensure the inherited host_memory_resource overload allocate(bytes) is usable
+TEST(PinnedMemoryResource, AllocateBytesOverload)
+{
+  rmm::mr::pinned_memory_resource mr;
+
+  void* ptr{nullptr};
+  EXPECT_NO_THROW(ptr = mr.allocate(128));
+  EXPECT_NE(nullptr, ptr);
+  EXPECT_TRUE(is_pinned_memory(ptr));
+  EXPECT_NO_THROW(mr.deallocate(ptr, 128));
+}
+
+}  // namespace rmm::test

